server/src/lib/model/db.ts

```diff
@@ -586,4 +586,71 @@ export default class DB {
       );
     }
   }
+
+  async getUserPoints(client_id: string, challenge: string) {
+    const [[row]] = await this.mysql.query(
+      `
+      SELECT (bonus_points + clip_point + vote_point) AS user
+      FROM (
+          SELECT bonus_points, clip_point, COUNT(votes.id) AS vote_point
+          FROM (
+              SELECT challenger.client_id, start_date, end_date, bonus_points, COUNT(clips.id) AS clip_point
+              FROM (
+                  SELECT user_clients.client_id,
+                      challenges.start_date AS start_date,
+                      TIMESTAMPADD(WEEK, 3, challenges.start_date) AS end_date,
+                      SUM(achievements.points) AS bonus_points
+                  FROM user_clients
+                  LEFT JOIN enroll ON user_clients.client_id = enroll.client_id
+                  LEFT JOIN teams ON enroll.team_id = teams.id
+                  LEFT JOIN challenges ON teams.challenge_id = challenges.id AND enroll.challenge_id = challenges.id
+                  LEFT JOIN earn ON user_clients.client_id = earn.client_id
+                  LEFT JOIN achievements ON earn.achievement_id = achievements.id AND challenges.id = achievements.challenge_id
+                  WHERE user_clients.client_id = ?
+                  AND challenges.url_token = ?
+                  AND earn.earned_at BETWEEN challenges.start_date AND TIMESTAMPADD(WEEK, 3, challenges.start_date)
+                  GROUP BY user_clients.client_id, start_date, end_date
+              ) challenger
+              LEFT JOIN clips ON challenger.client_id = clips.client_id AND clips.created_at BETWEEN start_date AND end_date
+              GROUP BY challenger.client_id, start_date, end_date, bonus_points
+          ) speaker
+          LEFT JOIN votes ON speaker.client_id = votes.client_id AND votes.created_at BETWEEN start_date AND end_date
+          GROUP BY speaker.client_id, speaker.bonus_points, speaker.clip_point
+      ) voter
+      `,
+      [client_id, challenge]
+    );
+    return row;
+  }
```

Comment:
Wow, this is a huge query. Nice work! I couldn't find any issues with it, but we should definitely test these functions out before the challenge start date.

We should also start to think about performance, and if we want to index anything here...

Lgtm!